// @effect/stream/mjs/internal/sink.mjs
var _a;
import * as Chunk from "@effect/data/Chunk";
import * as Context from "@effect/data/Context";
import * as Duration from "@effect/data/Duration";
import * as Either from "@effect/data/Either";
import { constTrue, dual, identity } from "@effect/data/Function";
import * as HashMap from "@effect/data/HashMap";
import * as HashSet from "@effect/data/HashSet";
import * as Option from "@effect/data/Option";
import * as ReadonlyArray from "@effect/data/ReadonlyArray";
import * as Cause from "@effect/io/Cause";
import * as Clock from "@effect/io/Clock";
import * as Effect from "@effect/io/Effect";
import * as Exit from "@effect/io/Exit";
import * as Hub from "@effect/io/Hub";
import * as Queue from "@effect/io/Queue";
import * as Ref from "@effect/io/Ref";
import * as channel from "@effect/stream/internal/channel";
import * as mergeDecision from "@effect/stream/internal/channel/mergeDecision";
import * as core from "@effect/stream/internal/core";
/** @internal */
const SinkSymbolKey = "@effect/stream/Sink";
/** @internal */
export const SinkTypeId = /*#__PURE__*/Symbol.for(SinkSymbolKey);
/** @internal */
const sinkVariance = {
  _R: _ => _,
  _E: _ => _,
  _In: _ => _,
  _L: _ => _,
  _Z: _ => _
};
/** @internal */
export class SinkImpl {
  constructor(channel) {
    this.channel = channel;
    this[_a] = sinkVariance;
  }
}
_a = SinkTypeId;
/** @internal */
export const as = /*#__PURE__*/dual(2, (self, z) => map(() => z)(self));
/** @internal */
export const collectAll = () => new SinkImpl(collectAllLoop(Chunk.empty()));
/** @internal */
const collectAllLoop = acc => core.readWithCause(chunk => collectAllLoop(Chunk.concat(chunk)(acc)), core.failCause, () => core.succeed(acc));
/** @internal */
export const collectAllN = n => {
  return flatMap(builder => foldUntil(builder, n, (chunk, input) => Chunk.append(input)(chunk)))(fromEffect(Effect.sync(() => Chunk.empty())));
};
/** @internal */
export const collectAllFrom = self => collectAllWhileWith(Chunk.empty(), constTrue, (chunk, z) => Chunk.append(z)(chunk))(self);
/** @internal */
export const collectAllToMap = (key, merge) => {
  return foldLeftChunks(HashMap.empty(), (map, chunk) => Chunk.reduce(map, (map, input) => {
    const k = key(input);
    const v = HashMap.has(k)(map) ? merge(HashMap.unsafeGet(k)(map), input) : input;
    return HashMap.set(k, v)(map);
  })(chunk));
};
/** @internal */
export const collectAllToMapN = (n, key, merge) => {
  return foldWeighted(HashMap.empty(), n, (acc, input) => HashMap.has(key(input))(acc) ? 0 : 1, (acc, input) => {
    const k = key(input);
    const v = HashMap.has(k)(acc) ? merge(HashMap.unsafeGet(k)(acc), input) : input;
    return HashMap.set(k, v)(acc);
  });
};
/** @internal */
export const collectAllToSet = () => foldLeftChunks(HashSet.empty(), (acc, chunk) => Chunk.reduce(acc, (acc, input) => HashSet.add(input)(acc))(chunk));
/** @internal */
export const collectAllToSetN = n => foldWeighted(HashSet.empty(), n, (acc, input) => HashSet.has(input)(acc) ? 0 : 1, (acc, input) => HashSet.add(input)(acc));
/** @internal */
export const collectAllUntil = p => {
  return map(tuple => tuple[0])(fold([Chunk.empty(), true], tuple => tuple[1], ([chunk, _], input) => [Chunk.append(input)(chunk), !p(input)]));
};
/** @internal */
export const collectAllUntilEffect = p => {
  return map(tuple => tuple[0])(foldEffect([Chunk.empty(), true], tuple => tuple[1], ([chunk, _], input) => Effect.map(bool => [Chunk.append(input)(chunk), !bool])(p(input))));
};
/** @internal */
export const collectAllWhile = predicate => fromChannel(collectAllWhileReader(predicate, Chunk.empty()));
/** @internal */
const collectAllWhileReader = (predicate, done) => core.readWith(input => {
  const [collected, leftovers] = ReadonlyArray.span(predicate)(Chunk.toReadonlyArray(input));
  if (leftovers.length === 0) {
    return collectAllWhileReader(predicate, Chunk.concat(Chunk.unsafeFromArray(collected))(done));
  }
  return channel.zipRight(core.succeed(Chunk.concat(Chunk.unsafeFromArray(collected))(done)))(core.write(Chunk.unsafeFromArray(leftovers)));
}, core.fail, () => core.succeed(done));
/** @internal */
export const collectAllWhileEffect = predicate => fromChannel(collectAllWhileEffectReader(predicate, Chunk.empty()));
/** @internal */
const collectAllWhileEffectReader = (predicate, done) => core.readWith(input => core.flatMap(collected => {
  const leftovers = Chunk.drop(collected.length)(input);
  if (Chunk.isEmpty(leftovers)) {
    return collectAllWhileEffectReader(predicate, Chunk.concat(collected)(done));
  }
  return channel.zipRight(core.succeed(Chunk.concat(collected)(done)))(core.write(leftovers));
})(core.fromEffect(Effect.takeWhile(predicate)(input))), core.fail, () => core.succeed(done));
/** @internal */
export const collectAllWhileWith = /*#__PURE__*/dual(4, (self, z, p, f) => {
  const refs = Effect.zip(Ref.make(false))(Ref.make(Chunk.empty()));
  const newChannel = core.flatMap(([leftoversRef, upstreamDoneRef]) => {
    const upstreamMarker = core.readWith(input => core.flatMap(() => upstreamMarker)(core.write(input)), core.fail, done => channel.as(done)(core.fromEffect(Ref.set(upstreamDoneRef, true))));
    return core.pipeTo(collectAllWhileWithLoop(self, leftoversRef, upstreamDoneRef, z, p, f))(core.pipeTo(channel.bufferChunk(leftoversRef))(upstreamMarker));
  })(core.fromEffect(refs));
  return new SinkImpl(newChannel);
});
/** @internal */
const collectAllWhileWithLoop = (self, leftoversRef, upstreamDoneRef, currentResult, p, f) => {
  return channel.foldChannel(core.fail, ([leftovers, doneValue]) => p(doneValue) ? core.flatMap(() => core.flatMap(upstreamDone => {
    const accumulatedResult = f(currentResult, doneValue);
    return upstreamDone ? channel.as(accumulatedResult)(core.write(Chunk.flatten(leftovers))) : collectAllWhileWithLoop(self, leftoversRef, upstreamDoneRef, accumulatedResult, p, f);
  })(core.fromEffect(Ref.get(upstreamDoneRef))))(core.fromEffect(Ref.set(leftoversRef, Chunk.flatten(leftovers)))) : channel.as(currentResult)(core.write(Chunk.flatten(leftovers))))(channel.doneCollect(toChannel(self)));
};
/** @internal */
export const collectLeftover = self => new SinkImpl(channel.map(([chunks, z]) => [z, Chunk.flatten(chunks)])(core.collectElements(toChannel(self))));
/** @internal */
export const contramap = /*#__PURE__*/dual(2, (self, f) => contramapChunks(Chunk.map(f))(self));
/** @internal */
export const contramapEffect = /*#__PURE__*/dual(2, (self, f) => contramapChunksEffect(Effect.forEach(f))(self));
/** @internal */
export const contramapChunks = /*#__PURE__*/dual(2, (self, f) => {
  const loop = core.readWith(chunk => core.flatMap(() => loop)(core.write(f(chunk))), core.fail, core.succeed);
  return new SinkImpl(core.pipeTo(toChannel(self))(loop));
});
/** @internal */
export const contramapChunksEffect = /*#__PURE__*/dual(2, (self, f) => {
  const loop = core.readWith(chunk => core.flatMap(() => loop)(core.flatMap(core.write)(core.fromEffect(f(chunk)))), core.fail, core.succeed);
  return new SinkImpl(channel.pipeToOrFail(toChannel(self))(loop));
});
/** @internal */
export const count = () => foldLeft(0, (s, _) => s + 1);
/** @internal */
export const die = defect => failCause(Cause.die(defect));
/** @internal */
export const dieMessage = message => failCause(Cause.die(Cause.RuntimeException(message)));
/** @internal */
export const dieSync = evaluate => failCauseSync(() => Cause.die(evaluate()));
/** @internal */
export const dimap = /*#__PURE__*/dual(3, (self, f, g) => map(g)(contramap(f)(self)));
/** @internal */
export const dimapEffect = /*#__PURE__*/dual(3, (self, f, g) => mapEffect(g)(contramapEffect(f)(self)));
/** @internal */
export const dimapChunks = /*#__PURE__*/dual(3, (self, f, g) => map(g)(contramapChunks(f)(self)));
/** @internal */
export const dimapChunksEffect = /*#__PURE__*/dual(3, (self, f, g) => mapEffect(g)(contramapChunksEffect(f)(self)));
/** @internal */
export const drain = () => new SinkImpl(channel.drain(channel.identityChannel()));
/** @internal */
export const drop = n => suspend(() => new SinkImpl(dropLoop(n)));
/** @internal */
const dropLoop = n => core.readWith(input => {
  const dropped = Chunk.drop(n)(input);
  const leftover = Math.max(n - input.length, 0);
  const more = Chunk.isEmpty(input) || leftover > 0;
  if (more) {
    return dropLoop(leftover);
  }
  return channel.zipRight(channel.identityChannel())(core.write(dropped));
}, core.fail, core.unit);
/** @internal */
export const dropUntil = predicate => new SinkImpl(channel.pipeToOrFail(toChannel(drop(1)))(toChannel(dropWhile(input => !predicate(input)))));
/** @internal */
export const dropUntilEffect = predicate => suspend(() => new SinkImpl(dropUntilEffectReader(predicate)));
/** @internal */
const dropUntilEffectReader = predicate => core.readWith(input => channel.unwrap(Effect.map(leftover => {
  const more = Chunk.isEmpty(leftover);
  return more ? dropUntilEffectReader(predicate) : channel.zipRight(channel.identityChannel())(core.write(leftover));
})(Effect.dropUntil(predicate)(input))), core.fail, core.unit);
/** @internal */
export const dropWhile = predicate => new SinkImpl(dropWhileReader(predicate));
/** @internal */
const dropWhileReader = predicate => core.readWith(input => {
  const out = Chunk.dropWhile(predicate)(input);
  if (Chunk.isEmpty(out)) {
    return dropWhileReader(predicate);
  }
  return channel.zipRight(channel.identityChannel())(core.write(out));
}, core.fail, core.succeedNow);
/** @internal */
export const dropWhileEffect = predicate => suspend(() => new SinkImpl(dropWhileEffectReader(predicate)));
/** @internal */
const dropWhileEffectReader = predicate => core.readWith(input => channel.unwrap(Effect.map(leftover => {
  const more = Chunk.isEmpty(leftover);
  return more ? dropWhileEffectReader(predicate) : channel.zipRight(channel.identityChannel())(core.write(leftover));
})(Effect.dropWhile(predicate)(input))), core.fail, core.unit);
/** @internal */
export const ensuring = /*#__PURE__*/dual(2, (self, finalizer) => new SinkImpl(channel.ensuring(finalizer)(toChannel(self))));
/** @internal */
export const ensuringWith = /*#__PURE__*/dual(2, (self, finalizer) => new SinkImpl(core.ensuringWith(finalizer)(toChannel(self))));
/** @internal */
export const context = () => fromEffect(Effect.context());
/** @internal */
export const contextWith = f => map(f)(context());
/** @internal */
export const contextWithEffect = f => mapEffect(f)(context());
/** @internal */
export const contextWithSink = f => new SinkImpl(channel.unwrap(Effect.contextWith(context => toChannel(f(context)))));
/** @internal */
export const every = predicate => fold(true, identity, (acc, input) => acc && predicate(input));
/** @internal */
export const fail = e => new SinkImpl(core.fail(e));
/** @internal */
export const failSync = evaluate => new SinkImpl(core.failSync(evaluate));
/** @internal */
export const failCause = cause => new SinkImpl(core.failCause(cause));
/** @internal */
export const failCauseSync = evaluate => new SinkImpl(core.failCauseSync(evaluate));
/** @internal */
export const filterInput = f => {
  return self => contramapChunks(Chunk.filter(f))(self);
};
/** @internal */
export const filterInputEffect = /*#__PURE__*/dual(2, (self, f) => contramapChunksEffect(Effect.filter(f))(self));
/** @internal */
export const findEffect = /*#__PURE__*/dual(2, (self, f) => {
  const newChannel = core.flatMap(([leftoversRef, upstreamDoneRef]) => {
    const upstreamMarker = core.readWith(input => core.flatMap(() => upstreamMarker)(core.write(input)), core.fail, done => channel.as(done)(core.fromEffect(Ref.set(upstreamDoneRef, true))));
    const loop = channel.foldChannel(core.fail, ([leftovers, doneValue]) => core.flatMap(satisfied => channel.zipRight(core.flatMap(upstreamDone => {
      if (satisfied) {
        return channel.as(Option.some(doneValue))(core.write(Chunk.flatten(leftovers)));
      }
      if (upstreamDone) {
        return channel.as(Option.none())(core.write(Chunk.flatten(leftovers)));
      }
      return loop;
    })(core.fromEffect(Ref.get(upstreamDoneRef))))(core.fromEffect(Ref.set(leftoversRef, Chunk.flatten(leftovers)))))(core.fromEffect(f(doneValue))))(core.collectElements(toChannel(self)));
    return core.pipeTo(loop)(core.pipeTo(channel.bufferChunk(leftoversRef))(upstreamMarker));
  })(core.fromEffect(Effect.zip(Ref.make(false))(Ref.make(Chunk.empty()))));
  return new SinkImpl(newChannel);
});
/** @internal */
export const fold = (s, contFn, f) => suspend(() => new SinkImpl(foldReader(s, contFn, f)));
/** @internal */
const foldReader = (s, contFn, f) => {
  if (!contFn(s)) {
    return core.succeedNow(s);
  }
  return core.readWith(input => {
    const [nextS, leftovers] = foldChunkSplit(s, input, contFn, f, 0, input.length);
    if (Chunk.isNonEmpty(leftovers)) {
      return channel.as(nextS)(core.write(leftovers));
    }
    return foldReader(nextS, contFn, f);
  }, core.fail, () => core.succeedNow(s));
};
/** @internal */
const foldChunkSplit = (s, chunk, contFn, f, index, length) => {
  if (index === length) {
    return [s, Chunk.empty()];
  }
  const s1 = f(s, Chunk.unsafeGet(index)(chunk));
  if (contFn(s1)) {
    return foldChunkSplit(s1, chunk, contFn, f, index + 1, length);
  }
  return [s1, Chunk.drop(index + 1)(chunk)];
};
/** @internal */
export const foldSink = /*#__PURE__*/dual(3, (self, onFailure, onSuccess) => {
  const newChannel = channel.foldChannel(error => toChannel(onFailure(error)), ([leftovers, z]) => core.suspend(() => {
    const leftoversRef = {
      ref: Chunk.filter(Chunk.isNonEmpty)(leftovers)
    };
    const refReader =
    // This cast is safe because of the L1 >: L <: In1 bound. It follows that
    // L <: In1 and therefore Chunk[L] can be safely cast to Chunk[In1].
    core.flatMap(chunk => channel.writeChunk(chunk))(core.sync(() => {
      const ref = leftoversRef.ref;
      leftoversRef.ref = Chunk.empty();
      return ref;
    }));
    const passthrough = channel.identityChannel();
    const continuationSink = core.pipeTo(toChannel(onSuccess(z)))(channel.zipRight(passthrough)(refReader));
    return core.flatMap(([newLeftovers, z1]) => channel.as(z1)(channel.zipRight(channel.writeChunk(newLeftovers))(core.flatMap(channel.writeChunk)(core.succeed(leftoversRef.ref)))))(core.collectElements(continuationSink));
  }))(core.collectElements(toChannel(self)));
  return new SinkImpl(newChannel);
});
/** @internal */
export const foldChunks = (s, contFn, f) => suspend(() => new SinkImpl(foldChunksReader(s, contFn, f)));
/** @internal */
const foldChunksReader = (s, contFn, f) => {
  if (!contFn(s)) {
    return core.succeedNow(s);
  }
  return core.readWith(input => foldChunksReader(f(s, input), contFn, f), core.fail, () => core.succeedNow(s));
};
/** @internal */
export const foldChunksEffect = (s, contFn, f) => suspend(() => new SinkImpl(foldChunksEffectReader(s, contFn, f)));
/** @internal */
const foldChunksEffectReader = (s, contFn, f) => {
  if (!contFn(s)) {
    return core.succeedNow(s);
  }
  return core.readWith(input => core.flatMap(s => foldChunksEffectReader(s, contFn, f))(core.fromEffect(f(s, input))), core.fail, () => core.succeedNow(s));
};
/** @internal */
export const foldEffect = (s, contFn, f) => suspend(() => new SinkImpl(foldEffectReader(s, contFn, f)));
/** @internal */
const foldEffectReader = (s, contFn, f) => {
  if (!contFn(s)) {
    return core.succeedNow(s);
  }
  return core.readWith(input => core.flatMap(([nextS, leftovers]) => Option.match(() => foldEffectReader(nextS, contFn, f), leftover => channel.as(nextS)(core.write(leftover)))(leftovers))(core.fromEffect(foldChunkSplitEffect(s, input, contFn, f))), core.fail, () => core.succeedNow(s));
};
/** @internal */
const foldChunkSplitEffect = (s, chunk, contFn, f) => foldChunkSplitEffectInternal(s, chunk, 0, chunk.length, contFn, f);
/** @internal */
const foldChunkSplitEffectInternal = (s, chunk, index, length, contFn, f) => {
  if (index === length) {
    return Effect.succeed([s, Option.none()]);
  }
  return Effect.flatMap(s1 => contFn(s1) ? foldChunkSplitEffectInternal(s1, chunk, index + 1, length, contFn, f) : Effect.succeed([s1, Option.some(Chunk.drop(index + 1)(chunk))]))(f(s, Chunk.unsafeGet(index)(chunk)));
};
/** @internal */
export const foldLeft = (s, f) => ignoreLeftover(fold(s, constTrue, f));
/** @internal */
export const foldLeftChunks = (s, f) => foldChunks(s, constTrue, f);
/** @internal */
export const foldLeftChunksEffect = (s, f) => ignoreLeftover(foldChunksEffect(s, constTrue, f));
/** @internal */
export const foldLeftEffect = (s, f) => foldEffect(s, constTrue, f);
/** @internal */
export const foldUntil = (s, max, f) => map(tuple => tuple[0])(fold([s, 0], tuple => tuple[1] < max, ([output, count], input) => [f(output, input), count + 1]));
/** @internal */
export const foldUntilEffect = (s, max, f) => map(tuple => tuple[0])(foldEffect([s, 0], tuple => tuple[1] < max, ([output, count], input) => Effect.map(s => [s, count + 1])(f(output, input))));
/** @internal */
export const foldWeighted = (s, max, costFn, f) => foldWeightedDecompose(s, max, costFn, Chunk.of, f);
/** @internal */
export const foldWeightedDecompose = (s, max, costFn, decompose, f) => suspend(() => new SinkImpl(foldWeightedDecomposeLoop(s, 0, false, max, costFn, decompose, f)));
/** @internal */
const foldWeightedDecomposeLoop = (s, cost, dirty, max, costFn, decompose, f) => core.readWith(input => {
  const [nextS, nextCost, nextDirty, leftovers] = foldWeightedDecomposeFold(input, 0, s, cost, dirty, max, costFn, decompose, f);
  if (Chunk.isNonEmpty(leftovers)) {
    return channel.zipRight(core.succeedNow(nextS))(core.write(leftovers));
  }
  if (cost > max) {
    return core.succeedNow(nextS);
  }
  return foldWeightedDecomposeLoop(nextS, nextCost, nextDirty, max, costFn, decompose, f);
}, core.fail, () => core.succeedNow(s));
/** @internal */
const foldWeightedDecomposeFold = (input, index, s, cost, dirty, max, costFn, decompose, f) => {
  if (index === input.length) {
    return [s, cost, dirty, Chunk.empty()];
  }
  const elem = Chunk.unsafeGet(index)(input);
  const total = cost + costFn(s, elem);
  if (total <= max) {
    return foldWeightedDecomposeFold(input, index + 1, f(s, elem), total, true, max, costFn, decompose, f);
  }
  const decomposed = decompose(elem);
  if (decomposed.length <= 1 && !dirty) {
    // If `elem` cannot be decomposed, we need to cross the `max` threshold. To
    // minimize "injury", we only allow this when we haven't added anything else
    // to the aggregate (dirty = false).
    return [f(s, elem), total, true, Chunk.drop(index + 1)(input)];
  }
  if (decomposed.length <= 1 && dirty) {
    // If the state is dirty and `elem` cannot be decomposed, we stop folding
    // and include `elem` in the leftovers.
    return [s, cost, dirty, Chunk.drop(index)(input)];
  }
  // `elem` got decomposed, so we will recurse with the decomposed elements pushed
  // into the chunk we're processing and see if we can aggregate further.
  const next = Chunk.concat(Chunk.drop(index + 1)(input))(decomposed);
  return foldWeightedDecomposeFold(next, 0, s, cost, dirty, max, costFn, decompose, f);
};
/** @internal */
export const foldWeightedDecomposeEffect = (s, max, costFn, decompose, f) => suspend(() => new SinkImpl(foldWeightedDecomposeEffectLoop(s, max, costFn, decompose, f, 0, false)));
/** @internal */
export const foldWeightedEffect = (s, max, costFn, f) => foldWeightedDecomposeEffect(s, max, costFn, input => Effect.succeed(Chunk.of(input)), f);
/** @internal */
const foldWeightedDecomposeEffectLoop = (s, max, costFn, decompose, f, cost, dirty) => core.readWith(input => core.flatMap(([nextS, nextCost, nextDirty, leftovers]) => {
  if (Chunk.isNonEmpty(leftovers)) {
    return channel.zipRight(core.succeedNow(nextS))(core.write(leftovers));
  }
  if (cost > max) {
    return core.succeedNow(nextS);
  }
  return foldWeightedDecomposeEffectLoop(nextS, max, costFn, decompose, f, nextCost, nextDirty);
})(core.fromEffect(foldWeightedDecomposeEffectFold(s, max, costFn, decompose, f, input, dirty, cost, 0))), core.fail, () => core.succeedNow(s));
/** @internal */
const foldWeightedDecomposeEffectFold = (s, max, costFn, decompose, f, input, dirty, cost, index) => {
  if (index === input.length) {
    return Effect.succeed([s, cost, dirty, Chunk.empty()]);
  }
  const elem = Chunk.unsafeGet(index)(input);
  return Effect.flatMap(total => {
    if (total <= max) {
      return Effect.flatMap(s => foldWeightedDecomposeEffectFold(s, max, costFn, decompose, f, input, true, total, index + 1))(f(s, elem));
    }
    return Effect.flatMap(decomposed => {
      if (decomposed.length <= 1 && !dirty) {
        // If `elem` cannot be decomposed, we need to cross the `max` threshold. To
        // minimize "injury", we only allow this when we haven't added anything else
        // to the aggregate (dirty = false).
        return Effect.map(s => [s, total, true, Chunk.drop(index + 1)(input)])(f(s, elem));
      }
      if (decomposed.length <= 1 && dirty) {
        // If the state is dirty and `elem` cannot be decomposed, we stop folding
        // and include `elem` in th leftovers.
        return Effect.succeed([s, cost, dirty, Chunk.drop(index)(input)]);
      }
      // `elem` got decomposed, so we will recurse with the decomposed elements pushed
      // into the chunk we're processing and see if we can aggregate further.
      const next = Chunk.concat(Chunk.drop(index + 1)(input))(decomposed);
      return foldWeightedDecomposeEffectFold(s, max, costFn, decompose, f, next, dirty, cost, 0);
    })(decompose(elem));
  })(Effect.map(newCost => cost + newCost)(costFn(s, elem)));
};
/** @internal */
export const flatMap = /*#__PURE__*/dual(2, (self, f) => foldSink(fail, f)(self));
/** @internal */
export const forEach = f => {
  const process = core.readWithCause(input => core.flatMap(() => process)(core.fromEffect(Effect.forEachDiscard(f)(input))), core.failCause, core.unit);
  return new SinkImpl(process);
};
/** @internal */
export const forEachChunk = f => {
  const process = core.readWithCause(input => core.flatMap(() => process)(core.fromEffect(f(input))), core.failCause, core.unit);
  return new SinkImpl(process);
};
/** @internal */
export const forEachWhile = f => {
  const process = core.readWithCause(input => forEachWhileReader(f, input, 0, input.length, process), core.failCause, core.unit);
  return new SinkImpl(process);
};
/** @internal */
const forEachWhileReader = (f, input, index, length, cont) => {
  if (index === length) {
    return cont;
  }
  return channel.catchAll(error => channel.zipRight(core.fail(error))(core.write(Chunk.drop(index)(input))))(core.flatMap(bool => bool ? forEachWhileReader(f, input, index + 1, length, cont) : core.write(Chunk.drop(index)(input)))(core.fromEffect(f(Chunk.unsafeGet(index)(input)))));
};
/** @internal */
export const forEachChunkWhile = f => {
  const reader = core.readWith(input => core.flatMap(cont => cont ? reader : core.unit())(core.fromEffect(f(input))), core.fail, core.unit);
  return new SinkImpl(reader);
};
/** @internal */
export const fromChannel = channel => new SinkImpl(channel);
/** @internal */
export const fromEffect = effect => new SinkImpl(core.fromEffect(effect));
/** @internal */
export const fromHub = hub => fromQueue(hub);
/** @internal */
export const fromHubWithShutdown = hub => fromQueueWithShutdown(hub);
/** @internal */
export const fromPush = push => new SinkImpl(channel.unwrapScoped(Effect.map(fromPushPull)(push)));
const fromPushPull = push => core.readWith(input => channel.foldChannel(([either, leftovers]) => Either.match(error => channel.zipRight(core.fail(error))(core.write(leftovers)), z => channel.zipRight(core.succeedNow(z))(core.write(leftovers)))(either), () => fromPushPull(push))(core.fromEffect(push(Option.some(input)))), core.fail, () => channel.foldChannel(([either, leftovers]) => Either.match(error => channel.zipRight(core.fail(error))(core.write(leftovers)), z => channel.zipRight(core.succeedNow(z))(core.write(leftovers)))(either), () => core.fromEffect(Effect.dieMessage("BUG: Sink.fromPush - please report an issue at https://github.com/Effect-TS/stream/issues")))(core.fromEffect(push(Option.none()))));
/** @internal */
export const fromQueue = queue => forEachChunk(input => Queue.offerAll(queue, input));
/** @internal */
export const fromQueueWithShutdown = queue => unwrapScoped(Effect.map(fromQueue)(Effect.acquireRelease(Effect.succeed(queue), Queue.shutdown)));
/** @internal */
export const head = () => fold(Option.none(), Option.isNone, (option, input) => Option.match(() => Option.some(input), () => option)(option));
/** @internal */
export const ignoreLeftover = self => new SinkImpl(channel.drain(toChannel(self)));
/** @internal */
export const last = () => foldLeft(Option.none(), (_, input) => Option.some(input));
/** @internal */
export const leftover = chunk => new SinkImpl(core.suspend(() => core.write(chunk)));
/** @internal */
export const log = message => fromEffect(Effect.log(message));
/** @internal */
export const logDebug = message => fromEffect(Effect.logDebug(message));
/** @internal */
export const logDebugCause = cause => fromEffect(Effect.logDebugCause(cause));
/** @internal */
export const logDebugCauseMessage = (message, cause) => fromEffect(Effect.logDebugCauseMessage(message, cause));
/** @internal */
export const logError = message => fromEffect(Effect.logError(message));
/** @internal */
export const logErrorCause = cause => fromEffect(Effect.logErrorCause(cause));
/** @internal */
export const logErrorCauseMessage = (message, cause) => fromEffect(Effect.logErrorCauseMessage(message, cause));
/** @internal */
export const logFatal = message => fromEffect(Effect.logFatal(message));
/** @internal */
export const logFatalCause = cause => fromEffect(Effect.logFatalCause(cause));
/** @internal */
export const logFatalCauseMessage = (message, cause) => fromEffect(Effect.logFatalCauseMessage(message, cause));
/** @internal */
export const logInfo = message => fromEffect(Effect.logInfo(message));
/** @internal */
export const logInfoCause = cause => fromEffect(Effect.logInfoCause(cause));
/** @internal */
export const logInfoCauseMessage = (message, cause) => fromEffect(Effect.logInfoCauseMessage(message, cause));
/** @internal */
export const logWarning = message => fromEffect(Effect.logWarning(message));
/** @internal */
export const logWarningCause = cause => fromEffect(Effect.logWarningCause(cause));
/** @internal */
export const logWarningCauseMessage = (message, cause) => fromEffect(Effect.logWarningCauseMessage(message, cause));
/** @internal */
export const logTrace = message => fromEffect(Effect.logTrace(message));
/** @internal */
export const logTraceCause = cause => fromEffect(Effect.logTraceCause(cause));
/** @internal */
export const logTraceCauseMessage = (message, cause) => fromEffect(Effect.logTraceCauseMessage(message, cause));
/** @internal */
export const map = /*#__PURE__*/dual(2, (self, f) => {
  return new SinkImpl(channel.map(f)(toChannel(self)));
});
/** @internal */
export const mapEffect = /*#__PURE__*/dual(2, (self, f) => new SinkImpl(channel.mapEffect(f)(toChannel(self))));
/** @internal */
export const mapError = /*#__PURE__*/dual(2, (self, f) => new SinkImpl(channel.mapError(f)(toChannel(self))));
/** @internal */
export const mapLeftover = /*#__PURE__*/dual(2, (self, f) => new SinkImpl(channel.mapOut(Chunk.map(f))(toChannel(self))));
/** @internal */
export const mkString = () => suspend(() => {
  const strings = [];
  return map(() => strings.join(""))(foldLeftChunks(void 0, (_, elems) => Chunk.forEach(elems, elem => {
    strings.push(String(elem));
  })));
});
/** @internal */
export const never = () => fromEffect(Effect.never());
/** @internal */
export const orElse = /*#__PURE__*/dual(2, (self, that) => new SinkImpl(channel.orElse(() => toChannel(that()))(toChannel(self))));
/** @internal */
export const provideContext = /*#__PURE__*/dual(2, (self, context) => new SinkImpl(core.provideContext(context)(toChannel(self))));
/** @internal */
export const race = /*#__PURE__*/dual(2, (self, that) => map(Either.merge)(raceBoth(that)(self)));
/** @internal */
export const raceBoth = /*#__PURE__*/dual(2, (self, that) => raceBothCapacity(self, that, 16));
/** @internal */
export const raceBothCapacity = /*#__PURE__*/dual(3, (self, that, capacity) => {
  return raceWithCapacity(self, that, selfDone => mergeDecision.Done(Effect.map(Either.left)(Effect.done(selfDone))), thatDone => mergeDecision.Done(Effect.map(Either.right)(Effect.done(thatDone))), capacity);
});
/** @internal */
export const raceWith = /*#__PURE__*/dual(4, (self, that, leftDone, rightDone) => raceWithCapacity(self, that, leftDone, rightDone, 16));
/** @internal */
export const raceWithCapacity = /*#__PURE__*/dual(5, (self, that, leftDone, rightDone, capacity) => {
  const scoped = Effect.gen(function* ($) {
    const hub = yield* $(Hub.bounded(capacity));
    const channel1 = yield* $(channel.fromHubScoped(hub));
    const channel2 = yield* $(channel.fromHubScoped(hub));
    const reader = channel.toHub(hub);
    const writer = channel.mergeWith(core.pipeTo(toChannel(that))(channel2), leftDone, rightDone)(core.pipeTo(toChannel(self))(channel1));
    const racedChannel = channel.mergeWith(writer, () => mergeDecision.Await(exit => Effect.done(exit)), done => mergeDecision.Done(Effect.done(done)))(reader);
    return new SinkImpl(racedChannel);
  });
  return unwrapScoped(scoped);
});
/** @internal */
export const refineOrDie = /*#__PURE__*/dual(2, (self, pf) => refineOrDieWith(pf, identity)(self));
/** @internal */
export const refineOrDieWith = /*#__PURE__*/dual(3, (self, pf, f) => {
  const newChannel = channel.catchAll(error => Option.match(() => core.failCauseSync(() => Cause.die(f(error))), core.fail)(pf(error)))(toChannel(self));
  return new SinkImpl(newChannel);
});
/** @internal */
export const service = tag => serviceWith(tag, identity);
/** @internal */
export const serviceWith = (tag, f) => fromEffect(Effect.map(tag, f));
/** @internal */
export const serviceWithEffect = (tag, f) => fromEffect(Effect.flatMap(tag, f));
/** @internal */
export const serviceWithSink = (tag, f) => new SinkImpl(channel.unwrap(Effect.map(tag, service => toChannel(f(service)))));
/** @internal */
export const some = predicate => fold(false, bool => !bool, (acc, input) => acc || predicate(input));
/** @internal */
export const splitWhere = /*#__PURE__*/dual(2, (self, f) => {
  const newChannel = core.flatMap(ref => core.flatMap(([leftovers, z]) => core.flatMap(leftover => channel.zipRight(core.succeed(z))(core.write(Chunk.concat(Chunk.flatten(leftovers))(leftover))))(core.fromEffect(Ref.get(ref))))(core.collectElements(channel.pipeToOrFail(toChannel(self))(splitWhereSplitter(false, ref, f)))))(core.fromEffect(Ref.make(Chunk.empty())));
  return new SinkImpl(newChannel);
});
/** @internal */
const splitWhereSplitter = (written, leftovers, f) => core.readWithCause(input => {
  if (Chunk.isEmpty(input)) {
    return splitWhereSplitter(written, leftovers, f);
  }
  if (written) {
    const index = indexWhere(input, f);
    if (index === -1) {
      return channel.zipRight(splitWhereSplitter(true, leftovers, f))(core.write(input));
    }
    const [left, right] = Chunk.splitAt(index)(input);
    return channel.zipRight(core.fromEffect(Ref.set(leftovers, right)))(core.write(left));
  }
  const index = indexWhere(input, f, 1);
  if (index === -1) {
    return channel.zipRight(splitWhereSplitter(true, leftovers, f))(core.write(input));
  }
  const [left, right] = Chunk.splitAt(Math.max(index, 1))(input);
  return channel.zipRight(core.fromEffect(Ref.set(leftovers, right)))(core.write(left));
}, core.failCause, core.succeed);
/** @internal */
const indexWhere = (self, predicate, from = 0) => {
  const iterator = self[Symbol.iterator]();
  let index = 0;
  let result = -1;
  let next;
  while (result < 0 && (next = iterator.next()) && !next.done) {
    const a = next.value;
    if (index >= from && predicate(a)) {
      result = index;
    }
    index = index + 1;
  }
  return result;
};
/** @internal */
export const succeed = z => new SinkImpl(core.succeed(z));
/** @internal */
export const sum = () => foldLeft(0, (a, b) => a + b);
/** @internal */
export const summarized = /*#__PURE__*/dual(3, (self, summary, f) => {
  const newChannel = core.flatMap(start => core.flatMap(done => channel.map(end => [done, f(start, end)])(core.fromEffect(summary)))(toChannel(self)))(core.fromEffect(summary));
  return new SinkImpl(newChannel);
});
/** @internal */
export const suspend = evaluate => new SinkImpl(core.suspend(() => toChannel(evaluate())));
/** @internal */
export const sync = evaluate => new SinkImpl(core.sync(evaluate));
/** @internal */
export const take = n => flatMap(acc => {
  const [taken, leftover] = Chunk.splitAt(n)(acc);
  return new SinkImpl(channel.zipRight(core.succeedNow(taken))(core.write(leftover)));
})(foldChunks(Chunk.empty(), chunk => chunk.length < n, (acc, chunk) => Chunk.concat(chunk)(acc)));
/** @internal */
export const timed = () => map(tuple => tuple[1])(withDuration(drain()));
/** @internal */
export const toChannel = self => Context.isGenericTag(self) ? toChannel(fromEffect(self)) : self.channel;
/** @internal */
export const unwrap = effect => new SinkImpl(channel.unwrap(Effect.map(sink => toChannel(sink))(effect)));
/** @internal */
export const unwrapScoped = effect => {
  return new SinkImpl(channel.unwrapScoped(Effect.map(sink => toChannel(sink))(effect)));
};
/** @internal */
export const withDuration = self => summarized(Clock.currentTimeMillis(), (start, end) => Duration.millis(end - start))(self);
/** @internal */
export const zip = /*#__PURE__*/dual(2, (self, that) => zipWith(that, (z, z2) => [z, z2])(self));
/** @internal */
export const zipLeft = /*#__PURE__*/dual(2, (self, that) => zipWith(that, (z, _) => z)(self));
/** @internal */
export const zipRight = /*#__PURE__*/dual(2, (self, that) => zipWith(that, (_, z2) => z2)(self));
/** @internal */
export const zipWith = /*#__PURE__*/dual(3, (self, that, f) => flatMap(z => map(z2 => f(z, z2))(that))(self));
/** @internal */
export const zipPar = /*#__PURE__*/dual(2, (self, that) => zipWithPar(that, (z, z2) => [z, z2])(self));
/** @internal */
export const zipParLeft = /*#__PURE__*/dual(2, (self, that) => zipWithPar(that, (z, _) => z)(self));
/** @internal */
export const zipParRight = /*#__PURE__*/dual(2, (self, that) => zipWithPar(that, (_, z2) => z2)(self));
/** @internal */
export const zipWithPar = /*#__PURE__*/dual(3, (self, that, f) => {
  return raceWith(that, Exit.match(cause => mergeDecision.Done(Effect.failCause(cause)), leftZ => mergeDecision.Await(Exit.match(Effect.failCause, rightZ => Effect.succeed(f(leftZ, rightZ))))), Exit.match(cause => mergeDecision.Done(Effect.failCause(cause)), rightZ => mergeDecision.Await(Exit.match(Effect.failCause, leftZ => Effect.succeed(f(leftZ, rightZ))))))(self);
});
// Circular with Channel
/** @internal */
export const channelToSink = self => new SinkImpl(self);
