// dfx/DiscordConfig.js
import * as tsplus_module_1 from "dfx/_common";
import * as tsplus_module_2 from "@effect/data/Duration";
import * as tsplus_module_3 from "@effect/io/Layer";
import * as tsplus_module_4 from "@effect/data/Function";
import * as tsplus_module_5 from "@effect/io/Config";
import * as tsplus_module_6 from "@effect/io/Effect";
const VERSION = 10;
export const DiscordConfig = tsplus_module_1.Tag();
export const make = ({ token, rest, gateway }) => ({
    token,
    rest: {
        baseUrl: `https://discord.com/api/v${VERSION}`,
        ...(rest ?? {}),
        globalRateLimit: {
            limit: 50,
            window: tsplus_module_2.seconds(1),
            ...(rest?.globalRateLimit ?? {}),
        },
    },
    gateway: {
        intents: tsplus_module_1.Discord.GatewayIntents.GUILDS,
        identifyRateLimit: [5000, 1],
        ...(gateway ?? {}),
    },
});
export const makeLayer = tsplus_module_4.flow(make, _ => tsplus_module_3.succeed(DiscordConfig, _));
export const makeFromConfig = (_) => tsplus_module_6.toLayer(tsplus_module_6.map(tsplus_module_6.config(tsplus_module_5.unwrap(_)), make), DiscordConfig);
