// dfx/webhooks.js
import * as tsplus_module_1 from "@effect/io/Layer";
import * as tsplus_module_2 from "@effect/io/Config";
import * as tsplus_module_3 from "dfx";
import { LiveFetchRequestExecutor } from "@effect-http/client";
import { LiveDiscordREST } from "./DiscordREST.js";
import { makeFromConfig } from "./Interactions/webhook.js";
import { LiveMemoryRateLimitStore, LiveRateLimiter } from "./RateLimit.js";
export { BadWebhookSignature, makeConfigLayer, makeHandler, makeSimpleHandler, WebhookConfig, WebhookParseError, } from "./Interactions/webhook.js";
export const MemoryRateLimit = tsplus_module_1.provide(LiveRateLimiter)(LiveMemoryRateLimitStore);
export const MemoryREST = tsplus_module_1.provide(LiveDiscordREST)(LiveMemoryRateLimitStore);
export const makeLiveWithoutFetch = (options, debug = false) => {
    const config = tsplus_module_2.unwrap(options);
    const LiveWebhook = makeFromConfig(config);
    const LiveLog = debug ? tsplus_module_3.Log.LiveLogDebug : tsplus_module_3.Log.LiveLog;
    const LiveConfig = tsplus_module_3.DiscordConfig.makeFromConfig(config);
    const LiveEnv = tsplus_module_1.provide((tsplus_module_1.merge(MemoryRateLimit)(tsplus_module_1.merge(LiveWebhook)(MemoryREST))))((tsplus_module_1.merge(LiveConfig)(LiveLog)));
    return LiveEnv;
};
export const makeLive = (config, debug = false) => {
    return tsplus_module_1.provide(makeLiveWithoutFetch(config, debug))(LiveFetchRequestExecutor);
};
