// @effect/stream/mjs/internal/stream/haltStrategy.mjs
import { dual } from "@effect/data/Function";
import * as OpCodes from "@effect/stream/internal/opCodes/haltStrategy";
/** @internal */
export const Left = {
  _tag: OpCodes.OP_LEFT
};
/** @internal */
export const Right = {
  _tag: OpCodes.OP_RIGHT
};
/** @internal */
export const Both = {
  _tag: OpCodes.OP_BOTH
};
/** @internal */
export const Either = {
  _tag: OpCodes.OP_EITHER
};
/** @internal */
export const isLeft = self => self._tag === OpCodes.OP_LEFT;
/** @internal */
export const isRight = self => self._tag === OpCodes.OP_RIGHT;
/** @internal */
export const isBoth = self => self._tag === OpCodes.OP_BOTH;
/** @internal */
export const isEither = self => self._tag === OpCodes.OP_EITHER;
/** @internal */
export const match = /*#__PURE__*/dual(5, (self, onLeft, onRight, onBoth, onEither) => {
  switch (self._tag) {
    case OpCodes.OP_LEFT:
      {
        return onLeft();
      }
    case OpCodes.OP_RIGHT:
      {
        return onRight();
      }
    case OpCodes.OP_BOTH:
      {
        return onBoth();
      }
    case OpCodes.OP_EITHER:
      {
        return onEither();
      }
  }
});
