// @effect/stream/internal/opCodes/channel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OP_TRACED = exports.OP_SUSPEND = exports.OP_SUCCEED_NOW = exports.OP_SUCCEED = exports.OP_READ = exports.OP_PROVIDE = exports.OP_PIPE_TO = exports.OP_FROM_EFFECT = exports.OP_FOLD = exports.OP_FAIL = exports.OP_ENSURING = exports.OP_EMIT = exports.OP_CONCAT_ALL = exports.OP_BRIDGE = exports.OP_BRACKET_OUT = void 0;
/** @internal */
const OP_BRACKET_OUT = "BracketOut";
/** @internal */
exports.OP_BRACKET_OUT = OP_BRACKET_OUT;
const OP_BRIDGE = "Bridge";
/** @internal */
exports.OP_BRIDGE = OP_BRIDGE;
const OP_CONCAT_ALL = "ConcatAll";
/** @internal */
exports.OP_CONCAT_ALL = OP_CONCAT_ALL;
const OP_EMIT = "Emit";
/** @internal */
exports.OP_EMIT = OP_EMIT;
const OP_ENSURING = "Ensuring";
/** @internal */
exports.OP_ENSURING = OP_ENSURING;
const OP_FAIL = "Fail";
/** @internal */
exports.OP_FAIL = OP_FAIL;
const OP_FOLD = "Fold";
/** @internal */
exports.OP_FOLD = OP_FOLD;
const OP_FROM_EFFECT = "FromEffect";
/** @internal */
exports.OP_FROM_EFFECT = OP_FROM_EFFECT;
const OP_PIPE_TO = "PipeTo";
/** @internal */
exports.OP_PIPE_TO = OP_PIPE_TO;
const OP_PROVIDE = "Provide";
/** @internal */
exports.OP_PROVIDE = OP_PROVIDE;
const OP_READ = "Read";
/** @internal */
exports.OP_READ = OP_READ;
const OP_SUCCEED = "Succeed";
/** @internal */
exports.OP_SUCCEED = OP_SUCCEED;
const OP_SUCCEED_NOW = "SucceedNow";
/** @internal */
exports.OP_SUCCEED_NOW = OP_SUCCEED_NOW;
const OP_SUSPEND = "Suspend";
/** @internal */
exports.OP_SUSPEND = OP_SUSPEND;
const OP_TRACED = "Traced";
exports.OP_TRACED = OP_TRACED;
