// dfx/Interactions/definitions.js
import * as tsplus_module_1 from "dfx/_common";
import { EffectTypeId } from "@effect/io/Effect";
export class GlobalApplicationCommand {
    command;
    handle;
    _tag = "GlobalApplicationCommand";
    constructor(command, handle) {
        this.command = command;
        this.handle = handle;
    }
}
export const global = (command, handle) => new GlobalApplicationCommand(command, handle);
export class GuildApplicationCommand {
    command;
    handle;
    _tag = "GuildApplicationCommand";
    constructor(command, handle) {
        this.command = command;
        this.handle = handle;
    }
}
export const guild = (command, handle) => new GuildApplicationCommand(command, handle);
export class MessageComponent {
    predicate;
    handle;
    _tag = "MessageComponent";
    constructor(predicate, handle) {
        this.predicate = predicate;
        this.handle = handle;
    }
}
export const messageComponent = (pred, handle) => new MessageComponent(pred, handle);
export class ModalSubmit {
    predicate;
    handle;
    _tag = "ModalSubmit";
    constructor(predicate, handle) {
        this.predicate = predicate;
        this.handle = handle;
    }
}
export const modalSubmit = (pred, handle) => new ModalSubmit(pred, handle);
export class Autocomplete {
    predicate;
    handle;
    _tag = "Autocomplete";
    constructor(predicate, handle) {
        this.predicate = predicate;
        this.handle = handle;
    }
}
export const autocomplete = (pred, handle) => new Autocomplete(pred, handle);
