// @effect/stream/mjs/internal/channel/channelState.mjs
import * as Effect from "@effect/io/Effect";
import * as OpCodes from "@effect/stream/internal/opCodes/channelState";
/** @internal */
export const ChannelStateTypeId = /*#__PURE__*/Symbol.for("@effect/stream/Channel/State");
/** @internal */
const channelStateVariance = {
  _R: _ => _,
  _E: _ => _
};
/** @internal */
const proto = {
  [ChannelStateTypeId]: channelStateVariance
};
/** @internal */
export const Done = () => {
  const op = Object.create(proto);
  op._tag = OpCodes.OP_DONE;
  return op;
};
/** @internal */
export const Emit = () => {
  const op = Object.create(proto);
  op._tag = OpCodes.OP_EMIT;
  return op;
};
/** @internal */
export const FromEffect = effect => {
  const op = Object.create(proto);
  op._tag = OpCodes.OP_FROM_EFFECT;
  op.effect = effect;
  return op;
};
/** @internal */
export const Read = (upstream, onEffect, onEmit, onDone) => {
  const op = Object.create(proto);
  op._tag = OpCodes.OP_READ;
  op.upstream = upstream;
  op.onEffect = onEffect;
  op.onEmit = onEmit;
  op.onDone = onDone;
  return op;
};
/** @internal */
export const isChannelState = u => typeof u === "object" && u != null && ChannelStateTypeId in u;
/** @internal */
export const isDone = self => self._tag === OpCodes.OP_DONE;
/** @internal */
export const isEmit = self => self._tag === OpCodes.OP_EMIT;
/** @internal */
export const isFromEffect = self => self._tag === OpCodes.OP_FROM_EFFECT;
/** @internal */
export const isRead = self => self._tag === OpCodes.OP_READ;
/** @internal */
export const effect = self => isFromEffect(self) ? self.effect : Effect.unit();
/** @internal */
export const effectOrUndefinedIgnored = self => isFromEffect(self) ? Effect.ignore(self.effect) : undefined;
