// @effect/stream/mjs/internal/channel/childExecutorDecision.mjs
import { dual } from "@effect/data/Function";
import * as OpCodes from "@effect/stream/internal/opCodes/childExecutorDecision";
/** @internal */
const ChildExecutorDecisionSymbolKey = "@effect/stream/Channel/ChildExecutorDecision";
/** @internal */
export const ChildExecutorDecisionTypeId = /*#__PURE__*/Symbol.for(ChildExecutorDecisionSymbolKey);
/** @internal */
const proto = {
  [ChildExecutorDecisionTypeId]: ChildExecutorDecisionTypeId
};
/** @internal */
export const Continue = _ => {
  const op = Object.create(proto);
  op._tag = OpCodes.OP_CONTINUE;
  return op;
};
/** @internal */
export const Close = value => {
  const op = Object.create(proto);
  op._tag = OpCodes.OP_CLOSE;
  op.value = value;
  return op;
};
/** @internal */
export const Yield = _ => {
  const op = Object.create(proto);
  op._tag = OpCodes.OP_YIELD;
  return op;
};
/** @internal */
export const isChildExecutorDecision = u => typeof u === "object" && u != null && ChildExecutorDecisionTypeId in u;
/** @internal */
export const isContinue = self => self._tag === OpCodes.OP_CONTINUE;
/** @internal */
export const isClose = self => self._tag === OpCodes.OP_CLOSE;
/** @internal */
export const isYield = self => self._tag === OpCodes.OP_YIELD;
/** @internal */
export const match = /*#__PURE__*/dual(4, (self, onContinue, onClose, onYield) => {
  switch (self._tag) {
    case OpCodes.OP_CONTINUE:
      {
        return onContinue();
      }
    case OpCodes.OP_CLOSE:
      {
        return onClose(self.value);
      }
    case OpCodes.OP_YIELD:
      {
        return onYield();
      }
  }
});
