// dfx/Helpers/ui.js
/**
 * Helper to create an Action Row grid.
 */
export const grid = (items) => items.map((components) => ({
    type: 1 /* ComponentType.ACTION_ROW */,
    components,
}));
/**
 * Helper to create a single column of components
 */
export const singleColumn = (items) => items.map(c => ({
    type: 1 /* ComponentType.ACTION_ROW */,
    components: [c],
}));
/**
 * Helper to create a button component.
 */
export const button = (button) => ({
    type: 2 /* ComponentType.BUTTON */,
    style: 1 /* ButtonStyle.PRIMARY */,
    ...button,
});
/**
 * Helper to create a select component.
 */
export const select = (select) => ({
    type: 3 /* ComponentType.STRING_SELECT */,
    ...select,
});
/**
 * Helper to create a select component.
 */
export const userSelect = (select) => ({
    type: 5 /* ComponentType.USER_SELECT */,
    ...select,
});
/**
 * Helper to create a select component.
 */
export const roleSelect = (select) => ({
    type: 6 /* ComponentType.ROLE_SELECT */,
    ...select,
});
/**
 * Helper to create a select component.
 */
export const mentionableSelect = (select) => ({
    type: 7 /* ComponentType.MENTIONABLE_SELECT */,
    ...select,
});
/**
 * Helper to create a select component.
 */
export const channelSelect = (select) => ({
    type: 8 /* ComponentType.CHANNEL_SELECT */,
    ...select,
});
/**
 * Helper to create a text input
 */
export const textInput = (input) => ({
    type: 4 /* ComponentType.TEXT_INPUT */,
    style: 1 /* TextInputStyle.SHORT */,
    ...input,
});
