// @effect/stream/mjs/SubscriptionRef.mjs
import * as Ref from "@effect/io/Ref";
import * as Synchronized from "@effect/io/Ref/Synchronized";
import * as internal from "@effect/stream/internal/subscriptionRef";
/**
 * @since 1.0.0
 * @category symbols
 */
export const SubscriptionRefTypeId = internal.SubscriptionRefTypeId;
/**
 * @macro traced
 * @since 1.0.0
 * @category getters
 */
export const get = internal.get;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const getAndSet = Ref.getAndSet;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdate = Ref.getAndUpdate;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdateEffect = Synchronized.getAndUpdateEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdateSome = Ref.getAndUpdateSome;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdateSomeEffect = Synchronized.getAndUpdateSomeEffect;
/**
 * Creates a new `SubscriptionRef` with the specified value.
 *
 * @macro traced
 * @since 1.0.0
 * @category constructors
 */
export const make = internal.make;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const modify = internal.modify;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const modifyEffect = internal.modifyEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const modifySome = Ref.modifySome;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const modifySomeEffect = Synchronized.modifySomeEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const set = internal.set;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const setAndGet = Ref.setAndGet;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const update = Ref.update;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const updateEffect = Synchronized.updateEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const updateAndGet = Ref.updateAndGet;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const updateAndGetEffect = Synchronized.updateAndGetEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const updateSome = Ref.updateSome;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const updateSomeEffect = Synchronized.updateSomeEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const updateSomeAndGet = Ref.updateSomeAndGet;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
export const updateSomeAndGetEffect = Synchronized.updateSomeAndGetEffect;
