// @effect/stream/Stream/HaltStrategy.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.match = exports.isRight = exports.isLeft = exports.isEither = exports.isBoth = exports.Right = exports.Left = exports.Either = exports.Both = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/stream/internal/stream/haltStrategy"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category constructors
 */
const Left = internal.Left;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.Left = Left;
const Right = internal.Right;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.Right = Right;
const Both = internal.Both;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.Both = Both;
const Either = internal.Either;
/**
 * @since 1.0.0
 * @category refinements
 */
exports.Either = Either;
const isLeft = internal.isLeft;
/**
 * @since 1.0.0
 * @category refinements
 */
exports.isLeft = isLeft;
const isRight = internal.isRight;
/**
 * @since 1.0.0
 * @category refinements
 */
exports.isRight = isRight;
const isBoth = internal.isBoth;
/**
 * @since 1.0.0
 * @category refinements
 */
exports.isBoth = isBoth;
const isEither = internal.isEither;
/**
 * @since 1.0.0
 * @category folding
 */
exports.isEither = isEither;
const match = internal.match;
exports.match = match;
