// @effect/stream/SubscriptionRef.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateSomeEffect = exports.updateSomeAndGetEffect = exports.updateSomeAndGet = exports.updateSome = exports.updateEffect = exports.updateAndGetEffect = exports.updateAndGet = exports.update = exports.setAndGet = exports.set = exports.modifySomeEffect = exports.modifySome = exports.modifyEffect = exports.modify = exports.make = exports.getAndUpdateSomeEffect = exports.getAndUpdateSome = exports.getAndUpdateEffect = exports.getAndUpdate = exports.getAndSet = exports.get = exports.SubscriptionRefTypeId = void 0;
var Ref = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/Ref"));
var Synchronized = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/Ref/Synchronized"));
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/stream/internal/subscriptionRef"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * @since 1.0.0
 * @category symbols
 */
const SubscriptionRefTypeId = internal.SubscriptionRefTypeId;
/**
 * @macro traced
 * @since 1.0.0
 * @category getters
 */
exports.SubscriptionRefTypeId = SubscriptionRefTypeId;
const get = internal.get;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.get = get;
const getAndSet = Ref.getAndSet;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.getAndSet = getAndSet;
const getAndUpdate = Ref.getAndUpdate;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.getAndUpdate = getAndUpdate;
const getAndUpdateEffect = Synchronized.getAndUpdateEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.getAndUpdateEffect = getAndUpdateEffect;
const getAndUpdateSome = Ref.getAndUpdateSome;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.getAndUpdateSome = getAndUpdateSome;
const getAndUpdateSomeEffect = Synchronized.getAndUpdateSomeEffect;
/**
 * Creates a new `SubscriptionRef` with the specified value.
 *
 * @macro traced
 * @since 1.0.0
 * @category constructors
 */
exports.getAndUpdateSomeEffect = getAndUpdateSomeEffect;
const make = internal.make;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.make = make;
const modify = internal.modify;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.modify = modify;
const modifyEffect = internal.modifyEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.modifyEffect = modifyEffect;
const modifySome = Ref.modifySome;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.modifySome = modifySome;
const modifySomeEffect = Synchronized.modifySomeEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.modifySomeEffect = modifySomeEffect;
const set = internal.set;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.set = set;
const setAndGet = Ref.setAndGet;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.setAndGet = setAndGet;
const update = Ref.update;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.update = update;
const updateEffect = Synchronized.updateEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.updateEffect = updateEffect;
const updateAndGet = Ref.updateAndGet;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.updateAndGet = updateAndGet;
const updateAndGetEffect = Synchronized.updateAndGetEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.updateAndGetEffect = updateAndGetEffect;
const updateSome = Ref.updateSome;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.updateSome = updateSome;
const updateSomeEffect = Synchronized.updateSomeEffect;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.updateSomeEffect = updateSomeEffect;
const updateSomeAndGet = Ref.updateSomeAndGet;
/**
 * @macro traced
 * @since 1.0.0
 * @category utils
 */
exports.updateSomeAndGet = updateSomeAndGet;
const updateSomeAndGetEffect = Synchronized.updateSomeAndGetEffect;
exports.updateSomeAndGetEffect = updateSomeAndGetEffect;
