// @effect/stream/Take.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tap = exports.of = exports.matchEffect = exports.match = exports.map = exports.make = exports.isSuccess = exports.isFailure = exports.isDone = exports.fromPull = exports.fromExit = exports.fromEffect = exports.failCause = exports.fail = exports.end = exports.done = exports.dieMessage = exports.die = exports.chunk = exports.TakeTypeId = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/stream/internal/take"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * @since 1.0.0
 * @category symbols
 */
const TakeTypeId = internal.TakeTypeId;
/**
 * Creates a `Take` with the specified chunk.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.TakeTypeId = TakeTypeId;
const chunk = internal.chunk;
/**
 * Creates a failing `Take` with the specified defect.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.chunk = chunk;
const die = internal.die;
/**
 * Creates a failing `Take` with the specified error message.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.die = die;
const dieMessage = internal.dieMessage;
/**
 * Transforms a `Take<E, A>` to an `Effect<never, E, A>`.
 *
 * @macro traced
 * @since 1.0.0
 * @category destructors
 */
exports.dieMessage = dieMessage;
const done = internal.done;
/**
 * Represents the end-of-stream marker.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.done = done;
const end = internal.end;
/**
 * Creates a failing `Take` with the specified error.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.end = end;
const fail = internal.fail;
/**
 * Creates a failing `Take` with the specified cause.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.fail = fail;
const failCause = internal.failCause;
/**
 * Creates an effect from `Effect<R, E, A>` that does not fail, but succeeds with
 * the `Take<E, A>`. Error from stream when pulling is converted to
 * `Take.failCause`. Creates a single value chunk.
 *
 * @macro traced
 * @since 1.0.0
 * @category constructors
 */
exports.failCause = failCause;
const fromEffect = internal.fromEffect;
/**
 * Creates a `Take` from an `Exit`.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.fromEffect = fromEffect;
const fromExit = internal.fromExit;
/**
 * Creates effect from `Effect<R, Option<E>, Chunk<A>>` that does not fail, but
 * succeeds with the `Take<E, A>`. Errors from stream when pulling are converted
 * to `Take.failCause`, and the end-of-stream is converted to `Take.end`.
 *
 * @macro traced
 * @since 1.0.0
 * @category constructors
 */
exports.fromExit = fromExit;
const fromPull = internal.fromPull;
/**
 * Checks if this `take` is done (`Take.end`).
 *
 * @since 1.0.0
 * @category getters
 */
exports.fromPull = fromPull;
const isDone = internal.isDone;
/**
 * Checks if this `take` is a failure.
 *
 * @since 1.0.0
 * @category getters
 */
exports.isDone = isDone;
const isFailure = internal.isFailure;
/**
 * Checks if this `take` is a success.
 *
 * @since 1.0.0
 * @category getters
 */
exports.isFailure = isFailure;
const isSuccess = internal.isSuccess;
/**
 * Constructs a `Take`.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.isSuccess = isSuccess;
const make = internal.make;
/**
 * Transforms `Take<E, A>` to `Take<E, B>` by applying function `f`.
 *
 * @since 1.0.0
 * @category mapping
 */
exports.make = make;
const map = internal.map;
/**
 * Folds over the failure cause, success value and end-of-stream marker to
 * yield a value.
 *
 * @since 1.0.0
 * @category destructors
 */
exports.map = map;
const match = internal.match;
/**
 * Effectful version of `Take.fold`.
 *
 * Folds over the failure cause, success value and end-of-stream marker to
 * yield an effect.
 *
 * @macro traced
 * @since 1.0.0
 * @category destructors
 */
exports.match = match;
const matchEffect = internal.matchEffect;
/**
 * Creates a `Take` with a single value chunk.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.matchEffect = matchEffect;
const of = internal.of;
/**
 * Returns an effect that effectfully "peeks" at the success of this take.
 *
 * @macro traced
 * @since 1.0.0
 * @category sequencing
 */
exports.of = of;
const tap = internal.tap;
exports.tap = tap;
