// @effect/stream/internal/stream/pull.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fromDequeue = exports.failCause = exports.fail = exports.end = exports.empty = exports.emitChunk = exports.emit = void 0;
var Chunk = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/data/Chunk"));
var _Debug = /*#__PURE__*/require("@effect/data/Debug");
var Option = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/data/Option"));
var Effect = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/Effect"));
var Queue = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/Queue"));
var take = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/stream/internal/take"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * @macro traced
 * @internal
 */
const emit = /*#__PURE__*/(0, _Debug.methodWithTrace)(trace => value => Effect.succeed(Chunk.of(value)).traced(trace));
/**
 * @macro traced
 * @internal
 */
exports.emit = emit;
const emitChunk = /*#__PURE__*/(0, _Debug.methodWithTrace)(trace => chunk => Effect.succeed(chunk).traced(trace));
/**
 * @macro traced
 * @internal
 */
exports.emitChunk = emitChunk;
const empty = /*#__PURE__*/(0, _Debug.methodWithTrace)(trace => () => Effect.succeed(Chunk.empty()).traced(trace));
/**
 * @macro traced
 * @internal
 */
exports.empty = empty;
const end = /*#__PURE__*/(0, _Debug.methodWithTrace)(trace => () => Effect.fail(Option.none()).traced(trace));
/**
 * @macro traced
 * @internal
 */
exports.end = end;
const fail = /*#__PURE__*/(0, _Debug.methodWithTrace)(trace => error => Effect.fail(Option.some(error)).traced(trace));
/**
 * @macro traced
 * @internal
 */
exports.fail = fail;
const failCause = /*#__PURE__*/(0, _Debug.methodWithTrace)(trace => cause => Effect.mapError(Option.some)(Effect.failCause(cause)).traced(trace));
/**
 * @macro traced
 * @internal
 */
exports.failCause = failCause;
const fromDequeue = /*#__PURE__*/(0, _Debug.methodWithTrace)(trace => dequeue => Effect.flatMap(take.done)(Queue.take(dequeue)).traced(trace));
exports.fromDequeue = fromDequeue;
