// @effect/stream/internal/stream/zipAllState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PullRight = exports.PullLeft = exports.PullBoth = exports.OP_PULL_RIGHT = exports.OP_PULL_LEFT = exports.OP_PULL_BOTH = exports.OP_DRAIN_RIGHT = exports.OP_DRAIN_LEFT = exports.DrainRight = exports.DrainLeft = void 0;
/** @internal */
const OP_DRAIN_LEFT = "DrainLeft";
/** @internal */
exports.OP_DRAIN_LEFT = OP_DRAIN_LEFT;
const OP_DRAIN_RIGHT = "DrainRight";
/** @internal */
exports.OP_DRAIN_RIGHT = OP_DRAIN_RIGHT;
const OP_PULL_BOTH = "PullBoth";
/** @internal */
exports.OP_PULL_BOTH = OP_PULL_BOTH;
const OP_PULL_LEFT = "PullLeft";
/** @internal */
exports.OP_PULL_LEFT = OP_PULL_LEFT;
const OP_PULL_RIGHT = "PullRight";
/** @internal */
exports.OP_PULL_RIGHT = OP_PULL_RIGHT;
const DrainLeft = {
  _tag: OP_DRAIN_LEFT
};
/** @internal */
exports.DrainLeft = DrainLeft;
const DrainRight = {
  _tag: OP_DRAIN_RIGHT
};
/** @internal */
exports.DrainRight = DrainRight;
const PullBoth = {
  _tag: OP_PULL_BOTH
};
/** @internal */
exports.PullBoth = PullBoth;
const PullLeft = rightChunk => ({
  _tag: OP_PULL_LEFT,
  rightChunk
});
/** @internal */
exports.PullLeft = PullLeft;
const PullRight = leftChunk => ({
  _tag: OP_PULL_RIGHT,
  leftChunk
});
exports.PullRight = PullRight;
